File: csharp/HDFSTools/HDFSClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.IO;
using Newtonsoft.Json;


namespace HDFSTools
{

    public class _FileStatus
    {
        public long accessTime;
        public long blockSize;
        public int childrenNum;
        public int fileId;
        public string group;
        public long length;
        public long modificationTime;
        public string owner;
        public string pathSuffix;
        public string permission;
        public int replication;
        public string type;
    }

    public class ListStatusResponse
    {
        public class _FileStatuses
        {
            public IList<_FileStatus> FileStatus;
        }
        public _FileStatuses FileStatuses;
    }

    public class FileStatusResponse
    {
        public _FileStatus FileStatus;
    }

    public class NameNode
    {
        private readonly string domain;
        private readonly int port;

        public NameNode(string domain, int port)
        {
            this.domain = domain;
            this.port = port;
        }

        public override string ToString()
        {
            return String.Format("{0}:{1}", this.domain, this.port);
        }
    }

    public class HDFSClient
    {
        private static NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

        private const int TimeOutInSecond = 10;
        private DateTime lastTimeRefreshActiveNameNode;
        private readonly WebAction webAction = new WebAction();


        private static class QueryUrl
        {
            public static readonly string NameNodeStatus = "http://{0}/jmx?qry=Hadoop:service=NameNode,name=NameNodeStatus";
            public static readonly string WebHdfs = "http://{{0}}/webhdfs/v1/{0}";
            public static readonly string MakeDirs = "?op=MKDIRS";
            public static readonly string Delete = "?op=DELETE&recursive={0}";
            public static readonly string Rename = "?op=RENAME&destination={0}";
            public static readonly string Create = "?op=CREATE&overwrite={0}";
            public static readonly string ListDir = "?op=LISTSTATUS";
            public static readonly string Open = "?op=OPEN";
            public static readonly string Status = "?op=GETFILESTATUS";
        }

        private static class FileType
        {
            public const string File = "FILE";
            public const string Directory = "DIRECTORY";
        }

        public class ListResult
        {
            public string[] Dirs { get; set; }
            public string[] Files { get; set; }

            public bool IsEmtpy()
            {
                return Dirs.Count() + Files.Count() == 0;
            }
        }

        private NameNode[] nns;
        private NameNode activeNameNode;

        private string CombineUrl(params string[] urls)
        {
            return string.Join("/", (from url in urls
                                     where !string.IsNullOrWhiteSpace(url)
                                     select url.Trim('/')));
        }

        /// <summary>
        /// </summary>
        /// <param name="nns">list of NameNode address</param>
        public HDFSClient(NameNode[] nns)
        {
            InitHDFSClient(nns);
        }

        /// <summary>
        /// </summary>
        /// <param name="nns">list of NameNode addres</param>
        private void InitHDFSClient(NameNode[] nns)
        {
            this.nns = nns;
            this.webAction.SetDefaultExceptionTrigger(ExceptionTrigger);
            RefreshActiveNameNode();
        }

        /// <summary>
        /// Looking for the active NameNode
        /// </summary>
        public void ExceptionTrigger(string url, WebException ex, bool isFinal)
        {
            if (!isFinal)
            {
                logger.Warn(String.Format("Can't send request: {0} with message {1}, try to refresh active HDFS NameNode",
                                          String.Format(url, this.activeNameNode),
                                          ex.Message));
                RefreshActiveNameNode();
            }
            else
            {
                logger.Error(String.Format("Can't send request: {0} with error message {1}",
                                          String.Format(url, this.activeNameNode),
                                          ex.Message));
            }
        }

        private bool TrySendHdfsRequest(string url, string method)
        {
            Func<string, string, bool> action = (string _url, string _method) =>
            {
                var response = SendHdfsRequest(_url, _method);
                response.Close();
                return true;
            };
            return webAction.Retry(url, method, action, ExceptionTrigger);
        }

        private HttpWebResponse SendHdfsRequest(string url, string method)
        {
            string requestUrl = FormatNameNode(url);

            return SendRequest(requestUrl, method);
        }

        private HttpWebResponse SendRequest(string url, string method)
        {
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = method;

            HttpWebResponse response = (HttpWebResponse)request.GetResponse();

            if (response.StatusCode != HttpStatusCode.OK)
            {
                string msg = String.Format("Receive status code: {0}", response.StatusCode);
                response.Close();
                throw new WebException(msg);
            }
            return response;
        }
        private string FormatNameNode(string url)
        {
            CheckForRefreshActiveNameNode();
            return String.Format(url, activeNameNode);
        }

        private void CheckForRefreshActiveNameNode()
        {
            TimeSpan ts = DateTime.UtcNow - this.lastTimeRefreshActiveNameNode;
            if (ts.TotalSeconds >= TimeOutInSecond)
            {
                RefreshActiveNameNode();
            }
        }

        public void RefreshActiveNameNode()
        {
            int index;
            for (index = 0; index < this.nns.Length; ++index)
            {
                var nameNode = this.nns[index];
                string url = String.Format(QueryUrl.NameNodeStatus, nameNode);
                Func<string, string, string> action = (string _url, string _method) =>
                {
                    var response = SendRequest(_url, _method);
                    StreamReader sr = new StreamReader(response.GetResponseStream());
                    string message = sr.ReadToEnd();
                    sr.Close();
                    response.Close();
                    return message;
                };

                string responseMessage = webAction.Retry(url, WebRequestMethods.Http.Get, action);
                if (responseMessage != null && responseMessage.Contains("active"))
                {
                    break;
                }
            }
            if (index == this.nns.Length)
            {
                throw new Exception("Can't find active HDFS NameNode");
            }

            this.activeNameNode = this.nns[index];
            if (index != 0)
            {
                this.nns[index] = this.nns[0];
                this.nns[0] = this.activeNameNode;
            }
            this.lastTimeRefreshActiveNameNode = DateTime.UtcNow;
        }

        public bool Makedirs(string path)
        {
            string url = String.Format(QueryUrl.WebHdfs, path.TrimStart('/')) + QueryUrl.MakeDirs;
            return TrySendHdfsRequest(url, WebRequestMethods.Http.Put);
        }

        public bool Rename(string src, string dst)
        {
            string url = String.Format(QueryUrl.WebHdfs, src.TrimStart('/')) + String.Format(QueryUrl.Rename, dst);
            return TrySendHdfsRequest(url, WebRequestMethods.Http.Put);
        }

        private bool Delete(string path, bool recursive = false)
        {
            string url = String.Format(QueryUrl.WebHdfs, path.TrimStart('/')) + String.Format(QueryUrl.Delete, recursive);
            return TrySendHdfsRequest(url, "DELETE");
        }

        private string GenerateTmpPath(string path)
        {
            string guid = Guid.NewGuid().ToString();
            return path + "_" + guid + "_tmp";
        }

        private bool UploadFile(string localPath, string remotePath)
        {
            string url = String.Format(QueryUrl.WebHdfs, remotePath.Trim('/')) + String.Format(QueryUrl.Create, true);

            Func<string, string, bool> action = (_url, _method) =>
            {
                _url = FormatNameNode(_url);
                using (WebClient wc = new WebClient())
                {
                    wc.UploadFile(FormatNameNode(_url), _method, localPath);
                }
                return true;
            };

            return webAction.Retry(url, WebRequestMethods.Http.Put, action, ExceptionTrigger);
        }

        private bool UploadByTmpFile(string localPath, string remotePath)
        {
            string remoteTmpPath = GenerateTmpPath(remotePath);
            bool success = UploadFile(localPath, remoteTmpPath);
            if (success)
            {
                bool renameSuccess = Rename(remoteTmpPath, remotePath);
                if (!renameSuccess || ListDir(remoteTmpPath) != null)
                {
                    Delete(remoteTmpPath);
                    return false;
                }
                return true;
            }
            else
            {
                Delete(remoteTmpPath);
                return false;
            }
        }

        private bool UploadDirectory(string localPath, string remotePath)
        {

            Queue<string> dirQueue = new Queue<string>();

            dirQueue.Enqueue("");
            while (dirQueue.Count != 0)
            {
                string pathSuffix = dirQueue.Dequeue();
                string rPath = CombineUrl(remotePath, pathSuffix);
                string lPath = Path.Combine(localPath, pathSuffix);

                Makedirs(rPath);

                foreach (var filePath in Directory.GetFiles(lPath))
                {
                    var fileName = Path.GetFileName(filePath);
                    UploadFile(filePath, CombineUrl(rPath, fileName));
                }
                foreach (var dirPath in Directory.GetDirectories(lPath))
                {
                    dirQueue.Enqueue(CombineUrl(pathSuffix, Path.GetFileName(dirPath)));
                }
            }

            return true;
        }

        private bool UploadByTmpDirectory(string localPath, string remotePath)
        {
            string remoteTmpPath = GenerateTmpPath(remotePath);
            bool success = UploadDirectory(localPath, remoteTmpPath);
            if (success)
            {
                if (GetStatus(remotePath) != null)
                {
                    Delete(remotePath, true);
                }

                bool renameSuccess = Rename(remoteTmpPath, remotePath);
                if (!renameSuccess || ListDir(remoteTmpPath) != null)
                {
                    Delete(remoteTmpPath);
                    return false;
                }
                return true;
            }
            else
            {
                Delete(remoteTmpPath);
                return false;
            }
        }

        public bool Upload(string localPath, string remotePath)
        {
            if (File.Exists(localPath))
            {
                return UploadByTmpFile(localPath, remotePath);
            }
            else if (Directory.Exists(localPath))
            {
                return UploadByTmpDirectory(localPath, remotePath);
            }
            else
            {
                logger.Warn(string.Format("Local path not exists: {0}", localPath));
                return false;
            }
        }

        /// <summary>
        /// List HDFS directory
        /// </summary>
        /// <param name="path"></param>
        /// <param name="errBoundary"></param>
        /// <returns>ListResult, null if path is not exists</returns>
        private ListResult ListDir(string path)
        {
            string url = String.Format(QueryUrl.WebHdfs, path.TrimStart('/')) + QueryUrl.ListDir;


            Func<string, string, ListStatusResponse> action = (string _url, string _method) =>
            {
                var response = SendHdfsRequest(_url, _method);
                StreamReader sr = new StreamReader(response.GetResponseStream());
                ListStatusResponse listStatus = JsonConvert.DeserializeObject<ListStatusResponse>(sr.ReadToEnd());
                sr.Close();
                response.Close();
                return listStatus;
            };
            var statuses = webAction.Retry(url, WebRequestMethods.Http.Get, action, ExceptionTrigger);

            if (statuses == null)
            {
                return null;
            }

            var listResult = new ListResult();

            listResult.Dirs = (from status in statuses.FileStatuses.FileStatus
                               where status.type == FileType.Directory && !string.IsNullOrEmpty(status.pathSuffix)
                               select status.pathSuffix).ToArray();

            listResult.Files = (from status in statuses.FileStatuses.FileStatus
                                where status.type == FileType.File && !string.IsNullOrEmpty(status.pathSuffix)
                                select status.pathSuffix).ToArray();
            return listResult;
        }

        private _FileStatus GetStatus(string path)
        {
            string url = String.Format(QueryUrl.WebHdfs, path.TrimStart('/')) + QueryUrl.Status;

            Func<string, string, _FileStatus> action = (_url, _method) =>
            {
                var response = SendHdfsRequest(_url, _method);
                StreamReader sr = new StreamReader(response.GetResponseStream());
                FileStatusResponse fsr = JsonConvert.DeserializeObject<FileStatusResponse>(sr.ReadToEnd());
                sr.Close();
                response.Close();
                return fsr.FileStatus;
            };

            return webAction.Retry(url, WebRequestMethods.Http.Get, action, ExceptionTrigger);
        }

        private bool DownloadFile(string remotePath, string localPath)
        {
            string url = String.Format(QueryUrl.WebHdfs, remotePath.TrimStart('/')) + QueryUrl.Open;

            Func<string, string, bool> action = (_url, _method) =>
            {
                using (WebClient wc = new WebClient())
                {
                    wc.DownloadFile(FormatNameNode(_url), localPath);
                }
                return true;
            };

            return webAction.Retry(url, null, action, ExceptionTrigger);
        }

        private bool DownloadDirectory(string remotePath, string localPath)
        {
            Queue<string> dirQueue = new Queue<string>();

            dirQueue.Enqueue("");
            while (dirQueue.Count != 0)
            {
                string pathSuffix = dirQueue.Dequeue();
                string rPath = CombineUrl(remotePath, pathSuffix);
                string lPath = CombineUrl(localPath, pathSuffix);

                Directory.CreateDirectory(lPath);

                ListResult listResult = ListDir(rPath);

                foreach (var fileName in listResult.Files)
                {
                    DownloadFile(CombineUrl(remotePath, pathSuffix, fileName),
                                 CombineUrl(localPath, pathSuffix, fileName));
                }

                foreach (var dirName in listResult.Dirs)
                {
                    dirQueue.Enqueue(CombineUrl(pathSuffix, dirName));
                }
            }
            return true;
        }

        private bool DownloadByTmpFile(string remotePath, string localPath)
        {
            string localTmpPath = GenerateTmpPath(localPath);
            bool success = DownloadFile(remotePath, localTmpPath);
            if (success)
            {
                CheckAndRemoveTargetData(localPath);
                File.Move(localTmpPath, localPath);
            }
            else
            {
                CheckAndRemoveTargetData(localTmpPath);
            }
            return success;
        }

        private bool DownloadByTmpDirectory(string remotePath, string localPath)
        {
            string localTmpPath = GenerateTmpPath(localPath);
            bool success = DownloadDirectory(remotePath, localTmpPath);
            if (success)
            {
                CheckAndRemoveTargetData(localPath);
                Directory.Move(localTmpPath, localPath);
            }
            else
            {
                CheckAndRemoveTargetData(localTmpPath);
            }
            return success;
        }

        private void CheckAndRemoveTargetData(string path)
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
            else if (Directory.Exists(path))
            {
                Directory.Delete(path, true);
            }
        }

        public bool Download(string remotePath, string localPath)
        {
            return Download(remotePath, localPath, true);
        }

        public bool Download(string remotePath, string localPath, bool overwrite)
        {

            if (File.Exists(localPath) || Directory.Exists(localPath))
            {
                logger.Warn(string.Format("Local path have been used: {0}", localPath));
                if (!overwrite)
                {
                    return false;
                }
            }

            _FileStatus fileStatus = GetStatus(remotePath);
            if (fileStatus == null)
            {
                logger.Warn("Remote path is not exists, please check");
                return false;
            }

            switch (fileStatus.type)
            {
                case FileType.File:
                    return DownloadByTmpFile(remotePath, localPath);
                case FileType.Directory:
                    return DownloadByTmpDirectory(remotePath, localPath);
                default:
                    logger.Error(string.Format("Unknow file type: {0}", fileStatus.type));
                    return false;
            }
        }
    }
}

File: python/hdfs_client/hdfs_client.py
import logging
import os
import shutil
import json
from os import path as os_path
from os.path import join as path_join
from uuid import uuid1
from time import time
from collections import namedtuple, deque

import requests
from requests import RequestException
from requests.compat import urljoin

from .exceptions import ActiveNamenodeNotFoundException
from .request_wrapper import RetryAction, ResultValue

logger = logging.getLogger(__name__)

NameNode = namedtuple('NameNode', ['host', 'port'])
HDFSData = namedtuple('HDFSData', ['dirs', 'files'])


class QueryURL(object):
    NAMENODE_STATUS = 'http://{host}:{port}/jmx?qry=Hadoop:service=NameNode,name=NameNodeStatus'
    WEBHDFS = 'http://{{host}}:{{port}}/webhdfs/v1/{0}'
    MAKEDIRS = '?op=MKDIRS'
    DELETE = '?op=DELETE&recursive={0}'
    RENAME = '?op=RENAME&destination={0}'
    CREATE = '?op=CREATE&overwrite={0}'
    LISTDIR = '?op=LISTSTATUS'
    OPEN = '?op=OPEN'
    STATUS = '?op=GETFILESTATUS'


class DataType(object):
    FILE = 'FILE'
    DIRECTORY = 'DIRECTORY'


def retry_request(exceptions, default_return_value=ResultValue.NOTHING):

    def outer_wrapper(func):

        def func_wrapper(ins, *args, **kwargs):
            proc = RetryAction(func, exceptions, default_return_value, error_trigger=ins.request_error_trigger)
            return proc(ins, *args, **kwargs)
        return func_wrapper

    return outer_wrapper


class HDFSClient(object):

    def __init__(self, namenodes):
        """Create an instance of HDFSClient for use

        Args:
          namenodes (iterable of `NameNode`): a sequence of NameNode objects.

        """
        if len(namenodes) == 0 or len(namenodes) > 2:
            raise ValueError('length of namenodes should in range(1, 2)')

        self._nns = namenodes
        self._refresh_active_namenode()

    def request_error_trigger(self, err, is_final):
        if is_final:
            logger.error('Can\'t send request, error message:\n{}'.format(str(err)))
        else:
            logger.warn('Can\'t send request, error message:\n{}, try to refresh \
                         active HDFS NameNode and send request again'.format(str(err)))
            self._refresh_active_namenode()

    @retry_request((RequestException, ActiveNamenodeNotFoundException))
    def _refresh_active_namenode(self):

        active_index = None

        for index, nn in enumerate(self._nns):
            url = QueryURL.NAMENODE_STATUS.format(host=nn.host, port=nn.port)
            response = requests.get(url)
            if response.ok and u'active' in response.content:
                active_index = index
                break

        if active_index is None:
            raise  ActiveNamenodeNotFoundException('Can\'t find active namenode')

        if active_index == 1:
            self._nns.reverse()

        self._active_namenode = self._nns[0]
        self._last_refresh_ts = time()

    def _format_namenode(self, url):
        return url.format(host=self._active_namenode.host,
                          port=self._active_namenode.port)

    def _send_request(self, url, request_method, **kwargs):
        request_url = self._format_namenode(url)
        response = request_method(request_url, **kwargs)
        return response

    def _send_simple_request(self, url, request_method):
        response = self._send_request(url, request_method)
        is_success = response.ok
        response.close()
        return is_success

    @retry_request(RequestException, False)
    def makedirs(self, path):
        url = urljoin(QueryURL.WEBHDFS.format(path.lstrip('/')),
                      QueryURL.MAKEDIRS)
        return self._send_simple_request(url, requests.put)

    @retry_request(RequestException, False)
    def rename(self, src, dst):
        url = urljoin(QueryURL.WEBHDFS.format(src.lstrip('/')),
                      QueryURL.RENAME.format(dst))
        return self._send_simple_request(url, requests.put)

    @retry_request(RequestException, False)
    def delete(self, path, recursive=False):
        url = urljoin(QueryURL.WEBHDFS.format(path.lstrip('/')),
                      QueryURL.DELETE.format(recursive))
        return self._send_simple_request(url, requests.delete)

    @retry_request(RequestException, None)
    def listdir(self, path):
        url = urljoin(QueryURL.WEBHDFS.format(path.lstrip('/')),
                      QueryURL.LISTDIR)

        response = self._send_request(url, requests.get)
        if response.ok:
            data_list = json.loads(response.content)[u'FileStatuses'][u'FileStatus']
            response.close()
            dirs = tuple(d[u'pathSuffix'] for d in data_list if d[u'type'] == DataType.DIRECTORY and d[u'pathSuffix'])
            files = tuple(f[u'pathSuffix'] for f in data_list if f[u'type'] == DataType.FILE and f[u'pathSuffix'])
            hdfs_data = HDFSData(dirs=dirs, files=files)
            return hdfs_data
        else:
            response.close()
            return None

    def _get_remote_data_type(self, path):
        url = urljoin(QueryURL.WEBHDFS.format(path.lstrip('/')),
                      QueryURL.STATUS)

        response = self._send_request(url, requests.get)
        if response.ok:
            status = json.loads(response.content)[u'FileStatus']
            response.close()
            return status[u'type']
        else:
            response.close()
            return None

    def _get_local_data_type(self, path):
        if os_path.exists(path):
            if os_path.isdir(path):
                return DataType.DIRECTORY
            elif os_path.isfile(path):
                return DataType.FILE
        return None

    def upload(self, local_path, remote_path, overwrite=False):
        if not overwrite and self._get_remote_data_type(remote_path) is not None:
            logger.error('Upload target path already exists: {}'.format(remote_path))
            return False

        data_type = self._get_local_data_type(local_path)
        if data_type is None:
            logger.error('Can\'t identify local data: {}'.format(local_path))
            return False
        else:
            return self._upload_by_tmp_data(local_path, remote_path, data_type)

    def _upload_by_tmp_data(self, local_path, remote_path, data_type):
        remote_tmp_path = self._generate_tmp_path(remote_path)

        is_success = self._upload_data(local_path, remote_tmp_path, data_type)

        if is_success:
            self._check_and_remove_remote_data(remote_path)
            self.rename(remote_tmp_path, remote_path)
        else:
            self._check_and_remove_remote_data(remote_tmp_path)

        return is_success

    def _upload_data(self, local_path, remote_path, data_type):
        if data_type == DataType.DIRECTORY:
            return self._upload_directory(local_path, remote_path)
        elif data_type == DataType.FILE:
            return self._upload_file(local_path, remote_path)
        else:
            logger.warn('Given type "{}" is not in option'.format(data_type))
            return False

    def _upload_directory(self, local_path, remote_path):
        is_success = True
        prefix_length = len(local_path) + 1
        self.makedirs(remote_path)

        for local_root, dirs, files in os.walk(local_path):
            remote_root = path_join(remote_path, local_root[prefix_length:])
            for d in dirs:
                is_success &= self.makedirs(path_join(remote_root, d))
            for f in files:
                is_success &= self._upload_file(path_join(local_root, f), path_join(remote_root, f))
        return is_success

    @retry_request(RequestException, False)
    def _upload_file(self, local_path, remote_path):
        is_success = False
        url = urljoin(QueryURL.WEBHDFS.format(remote_path.lstrip('/')),
                      QueryURL.CREATE.format(True))

        with open(local_path, 'rb') as f:
            file_name = os_path.basename(remote_path)
            response = self._send_request(url, requests.put, files={file_name: f})
            is_success = response.ok
            response.close()

        return is_success

    def download(self, remote_path, local_path, overwrite=False):
        if not overwrite and os_path.exists(local_path):
            logger.error('Download target path already exists: {}'.format(local_path))
            return False

        data_type = self._get_remote_data_type(remote_path)
        if data_type is None:
            logger.error('Can\'t query HDFS file type, path: {}'.format(remote_path))
            return False
        else:
            return self._download_by_tmp_data(remote_path, local_path, data_type)

    def _download_by_tmp_data(self, remote_path, local_path, data_type):
        local_tmp_path = self._generate_tmp_path(local_path)

        is_success = self._download_data(remote_path, local_tmp_path, data_type)

        if is_success:
            self._check_and_remove_local_data(local_path)
            os.rename(local_tmp_path, local_path)
        else:
            self._check_and_remove_data(local_tmp_path)

        return is_success

    def _download_data(self, remote_path, local_path, data_type):
        if data_type == DataType.DIRECTORY:
            return self._download_directory(remote_path, local_path)
        elif data_type == DataType.FILE:
            return self._download_file(remote_path, local_path)
        else:
            logger.warn('Given type "{}" is not in option'.format(data_type))
            return False

    def _download_directory(self, remote_path, local_path):
        is_success = True
        dir_queue = deque([''])
        while len(dir_queue) and is_success:
            current_dir = dir_queue.popleft()
            remote_dir_path = path_join(remote_path, current_dir)
            local_dir_path = path_join(local_path, current_dir)
            os.makedirs(local_dir_path)

            data_status = self.listdir(remote_dir_path)
            for d in data_status.dirs:
                dir_queue.append(path_join(current_dir, d))
            for f in data_status.files:
                is_success &= self._download_file(path_join(remote_dir_path, f),
                                                  path_join(local_dir_path, f))
        return is_success

    @retry_request((RequestException, IOError), False)
    def _download_file(self, remote_path, local_path):
        url = urljoin(QueryURL.WEBHDFS.format(remote_path.lstrip('/')),
                      QueryURL.OPEN)

        response = self._send_request(url, requests.get)
        is_success = response.ok
        if is_success:
            self._write_content(response, local_path)

        response.close()
        return is_success

    def _write_content(self, response, path):
        with open(path, 'wb') as f:
            for chunk in response:
                f.write(chunk)

    def _check_and_remove_local_data(self, path):
        if os_path.exists(path):
            if os_path.isfile(path):
                os.remove(path)
            elif os_path.isdir(path):
                shutil.rmtree(path)

    def _check_and_remove_remote_data(self, path):
        if self._get_remote_data_type(path) is not None:
            self.delete(path, True)

    def _generate_tmp_path(self, path):
        uuid = str(uuid1()).replace('-', '')
        return '{path}_{uuid}_tmp'.format(path=path, uuid=uuid)
